Feature parameter names (#169)
NixOS/rfcs: rfcs/0169-feature-parameter-names.md

related-issues: (will contain links to implementation PRs)
---

# Summary
[summary]: #summary

Nixpkgs has mechanisms to exports compile-time options of the upstream build
system through the named parameters of derivation function, but the names of
these parameters are inconsistent from package to package and from feature to
feature.

# Motivation
[motivation]: #motivation

Developer interface for building environment with specific features enabled and
disabled is more complicated and requires more knowledge of implementation
details compared to already existing systems.

For example, Gentoo Linux has exhaustive list of [all known configuration
flags](https://www.gentoo.org/support/use-flags) and has means to
programmatically query what flags are available for particular package. Nixpkgs
has neither, and this RFC strives to rectify it.

# Detailed design
[design]: #detailed-design

When upstream build system has options to enable or disable a optional features
or configure some compile-time parameter, like path to default mailbox or
initial size of hash map, derivation function (for short, package) MAY expose
compile-time options via feature parameters that match

> `^(with|conf|enable)_[a-z_0-9]$` regular expression. For short, we will refer

Reviewer: I'm open to using underscores in these flags, but I think it would require changes to https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#syntax, which says

> Use `lowerCamelCase` for variable names, not `UpperCamelCase`. Note, this rule does not apply to package attribute names, which instead follow the rules in [package naming](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-naming).

Author: Thank you for bringing it up. I'll add into the migration section step of updating this documentation.

---
feature: feature_parameter_names
start-date: 2024-01-10
author: Dmitry Bogatov
co-authors: (find a buddy later to help out with the RFC)
shepherd-team: (names, to be nominated and accepted by RFC steering committee)
shepherd-leader: (name to be appointed by RFC steering committee)
related-issues: (will contain links to implementation PRs)
---

# Summary
[summary]: #summary

Nixpkgs has mechanisms to exports compile-time options of the upstream build
system through the named parameters of derivation function, but the names of
these parameters are inconsistent from package to package and from feature to
feature.

# Motivation
[motivation]: #motivation

Developer interface for building environment with specific features enabled and
disabled is more complicated and requires more knowledge of implementation
details compared to already existing systems.

For example, Gentoo Linux has exhaustive list of [all known configuration
flags](https://www.gentoo.org/support/use-flags) and has means to
programmatically query what flags are available for particular package. Nixpkgs

> has neither, and this RFC strives to rectify it.

Reviewer 2: You seem to describe the problem of `.override` attributes not being easily discoverable, but the RFC doesn't actually address that at all from what I understand.

If that's the actual problem you're trying to solve, the easiest solution might be to make those attributes more easily discoverable, e.g. by rendering attributes with their documentation on search.nixos.org. Or on the deep end, something like https://discourse.nixos.org/t/working-group-member-search-module-system-for-packages/26574

However, the actual problem you seem to want to solve with this RFC is described better in your comment here: https://github.com/NixOS/nixpkgs/pull/234463#issuecomment-1574892207. If that's indeed the case, the motivation section should be rewritten to talk about that.

Author: It actually does address, since now feature parameters are clearly namespaced from all other. To make it clearer,
I rewrote the motivation part as list of good things that we will get by implementing this RFC. Thoughts?

these parameters are inconsistent from package to package and from feature to
feature.

# Motivation
[motivation]: #motivation

Developer interface for building environment with specific features enabled and
disabled is more complicated and requires more knowledge of implementation
details compared to already existing systems.

For example, Gentoo Linux has exhaustive list of [all known configuration
flags](https://www.gentoo.org/support/use-flags) and has means to
programmatically query what flags are available for particular package. Nixpkgs
has neither, and this RFC strives to rectify it.

# Detailed design
[design]: #detailed-design

When upstream build system has options to enable or disable a optional features
or configure some compile-time parameter, like path to default mailbox or
initial size of hash map, derivation function (for short, package) MAY expose
compile-time options via feature parameters that match
`^(with|conf|enable)_[a-z_0-9]$` regular expression. For short, we will refer
to them as `with`-parameter, `conf`-parameter and `enable`-parameter correspondingly,
collectively as `feature parameters`.

> Note that this RFC does not establish any requires whether package should or

Reviewer 3: ```suggestion
Note that this RFC does not establish any requirements whether package should or
```

Author: Thanks. Fixed.

1. Knowing that package has a feature parameter (out of knowning the upstream)
   does not mechanically translates to the name of feature parameter, it can be
   named in multiple ways.

2. Writing generic code is complicated. Even if list of packages all have
   optional dependency on package `foo`, duck-typing like

```
map (p: p.override { with_foo = false; }) packages
```

won't work because of lack of uniformity, leading ugly code downstream.

3. If user wants to globally disable support for optional feature or build
   dependency as much as possible, he needs to set multiple feature parameters,
   and he still can miss couple. E.g, it is not obvious whether the following
   list is enough to cover everything X11-related:

```
withX
withX11
x11Support
enableX
```

4. There is no clear separation between feature parameters and other arguments
   to the package. With naming mandated by this RFC, distinction is clear to
   cater to automatic queries.

5. As side effect of the proposad migration plan is compiling exahustive list
   of known feature parameters. It is not huge stretch to have CI check to
   enforce that this list is current, bringing nixpkgs on par with Gentoo

>    [use-flags](https://www.gentoo.org/support/use-flags).

Reviewer 3: Doesn't Gentoo separately invest effort in consistently providing the popular USE flags for almost all the  packages where they make any amount of sense?

Author: At some point did some manual comparison between Gentoo and Nixpkgs and learned that Nixpkgs exports no less of feature parameters than Gentoo, just in less uniform way. In other words, I expected Gentoo configuration to be more granular than Nixpkgs, but that wasn't the case.

I'll move the last sentence into the future work to not inadvertently imply that RFC demands Nixpkgs to have as much feature parameters as Gentoo has use flags.

Reviewer 3: My expectation was that Nixpkgs has a comparable number of feature parameters, but the corresponding feature parameters are provided by fewer packages, isn't it so?

Author: Probably. Many packages do something like

```
buildInputs = lib.optionals stdenv.isLinux [ foo ];
```

without bothering to introduce `with_foo` flag.

---
feature: feature_parameter_names
start-date: 2024-01-10
author: Dmitry Bogatov
co-authors: (find a buddy later to help out with the RFC)
shepherd-team: (names, to be nominated and accepted by RFC steering committee)
shepherd-leader: (name to be appointed by RFC steering committee)
related-issues: (will contain links to implementation PRs)
---

# Summary
[summary]: #summary

Nixpkgs has mechanisms to exports compile-time options of the upstream build
system through the named parameters of derivation function, but the names of
these parameters are inconsistent from package to package and from feature to
feature.

# Motivation
[motivation]: #motivation

Developer interface for building environment with specific features enabled and
disabled is more complicated and and involves more "you just have to know it"
compared to already existing systems.

> 1. Knowing that package has a feature parameter (out of knowning the upstream)

Reviewer 4: ```suggestion
1. Knowing that package has a feature parameter (out of knowing the upstream)
```

   does not mechanically translates to the name of feature parameter, it can be

>    named in multiple ways.

Reviewer 2: Doesn't `.overrideAttrs (old: { cmakeFlags = old.cmakeFlags or [] ++ [ "..." ] })` already do this in a fairly standard way? I don't think there's even a need to expose cmake flags further.

Author: Not everything uses cmake. There are multiple build systems around, and projects even switch them sometimes. And each project names corresponding flag slightly differently.

The whole point of distribution is to abstract away idiosyncrasies of the upstream. If we say that anybody who wants non-default configuration has to do their own research, a lot of effort will be duplicated.

Reviewer 2: Then I'm not sure if the motivational point of being able to re-use upstream flags really works out.

If you abstract over something, you need to deal with changes when they happen underneath. This works out if upstream never changes any build flags, but we cannot guarantee that. And different build systems might not even support the same naming conventions.

So if upstream ever renames a flag, say from `with_foo` to `with_foo2`, to make the abstraction work, the Nix code has to implement support for both flags and deprecate the older one. Otherwise you risk breaking users `.override`'s.

While this is doable, I don't think package authors should be bothered with this. If the upstream flag changes, the package implementation should just transparently switch to the new name without requiring any changes at `.override` call sites.

So this is why I think this motivational point should be dropped: We shouldn't be required to re-use upstream flags.

Author: I see your point. I'll reword.

My idea is that if I happen to know that emacs has optional dependency on alsa, and I happend to know
that it is exported by nixpkgs, than I can write this:

```nix
emacs.override { with_alsa = false }
```

No need to remember whether it is `supportAlsa` or `withAlsaLib`.

Reviewer 2: How about establishing `.override { alsa = null; }` as a pattern for that instead? This way we save a flag.

In the implementation this would look like

```nix
{
  stdenv,
  alsa,
}:
stdenv.mkDerivation {
  
  configureFlags = [
    (lib.optional (alsa != null) "--with-alsa")

    # Or with a more convenient function
    (lib.withPackageFeature alsa "alsa")
  ];

}
```

Reviewer 5: This is what we used to do in Nixpkgs but I'm not a fan of this. A couple of reasons off the top of my head:

* it only applies to `with` parameters. `enable` or `conf` parameters cannot be covered by this as they aren't usually connected to any one package
* You cannot sensibly disable support for something by default; you'd have to override the thing with null at callPackage site. Apply the "ffmpeg test" if you want to see what I mean.
* Conflates existance of the dep with enablement of dep
* Overriding a dep with null could cause exceptions in the drv because it assumes the dep is an attrset (e.g. `versionAtLeast dep.version "1.0"` would throw if dep was null)

It'd be great if we could have these reasons (and likely more) to the RFC text as an alternative and for historical context as this was the status quo at some point.

Author: I'll include your nice list into description of `alsa = null` alternative, probably this weekend. Thank you.

Author: Updated RFC to mention `foo ? null` pattern.

2. Writing generic code is complicated. Even if list of packages all have
   optional dependency on package `foo`, duck-typing like

```
map (p: p.override { with_foo = false; }) packages
```

> won't work because of lack of uniformity, leading ugly code downstream.

Reviewer 2: Shouldn't this also work with `.overrideAttrs (old: { cmakeFlags = old.cmakeFlags or [] ++ [ "..." ] })`? Of course this only works when all packages actually have this option.

Author: And use the same build system. And name their flags in the same way.

I wish upstreams were to do the work that this RFC tries to do, but it didn't happened in 1991, and it won't happen now.

3. If user wants to globally disable support for optional feature or build
   dependency as much as possible, he needs to set multiple feature parameters,
   and he still can miss couple. E.g, it is not obvious whether the following
   list is enough to cover everything X11-related:

```
withX
withX11
x11Support
enableX

> ```

Reviewer 2: It sounds like this would be fixed with a [`config` option](https://github.com/NixOS/nixpkgs/blob/20d0954713ffa76a48507e90db6ad558f1238cf6/pkgs/top-level/config.nix), which is for flags that are generic/useful enough to apply to many packages.

Author: Overrides on individual packages become problematic (discussed in alternatives). Any way, we need some guidelines what exactly name we should to into `config`.

Reviewer 2: Makes sense.

I think a convention to reuse the same `config` option name would be good, e.g. like this:

```nix
{
  config,
  withX11 ? config.withX11,
}:
...
```

Then we have consistency between per-package flags and global flags.

Author: I don't think we will need `config` for that. If you want global, just `overlays = [self: super: { with_x11 = false; }`

4. There is no clear separation between feature parameters and other arguments
   to the package. With naming mandated by this RFC, distinction is clear to
   cater to automatic queries.

> 5. Side effect of the proposdd migration plan is compiling of almost

Reviewer 4: ```suggestion
5. Side effect of the proposed migration plan is compiling of almost
```

   exahustive list of known feature parameters.

# Detailed design
[design]: #detailed-design

When upstream build system has options to enable or disable a optional features
or configure some compile-time parameter, like path to default mailbox or
initial size of hash map, derivation function (for short, package) MAY expose
compile-time options via feature parameters that match
`^(with|conf|enable)_[a-z_0-9]$` regular expression. For short, we will refer
to them as `with`-parameter, `conf`-parameter and `enable`-parameter correspondingly,
collectively as `feature parameters`.

Note that this RFC does not establish any requirements whether package should or

> should not have any feature parameters, this decision is left at descretion of

Reviewer 4: ```suggestion
should not have any feature parameters, this decision is left at discretion of
```

<details><summary>Example</summary>

```
{ lib, stdenv, mpfr, with_mpfr ? true }:

stdenv.mkDerivation {
  # ... snip ...
  configureFlags = lib.optionals with_mpfr ["--with-mpfr"];
  buildInputs = lib.optionals with_mpfr [ mpfr ];
  # ... snip ...
}
```

</details>

As special provision, if optional vendored dependency is exposed by feature
parameter, it MUST be `with`-parameter.

If feature parameter does not incur extra runtime dependencies, corresponding
feature parameter MUST be `enable`-parameter and its name SHOULD correspond to
the upstream naming, unless is causes major inconsistency with other packages.
Consistency of feature parameters across nixpkgs is more important than
matching to the names used by particular upstream. What constitutes a major
inconsistency is left at discretion of the package maintainer(s).

<details><summary>Example</summary>

If upstream uses `--enable-translations` to denote support for NLS,
corresponding feature parameter SHOULD be `enable_nls`, since it is much more
common term for enabling programs to produce output in non-English languages.

</details>

<details><summary>Rationale</summary>

> This distinction between `with` and `enable` feature parameters is based on Autoconf naming conventions.

Reviewer 4: ```suggestion
This distinction between `with` and `enable` feature parameters is based on Autotools naming conventions.
```

[Autoconf ENABLE](https://www.gnu.org/software/autoconf/manual/autoconf-2.66/html_node/Package-Options.html)

> [Autoconf WITH](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.72/html_node/External-Software.html)

Reviewer 4: You are using distinct versions of autotools here.

}:

# Assertions are optional and might be infeasible when package has huge amount
# of feature flags, but in general improves user experience.
assert enable_ssl -> with_openssl || with_wolfssl;

stdenv.mkDerivation {
  # ... snip ...
  configureFlags = lib.optionals enable_ssl ["--enable-ssl"];
  buildInputs = lib.optionals with_wolfssl [ wolfssl ]
             ++ lib.optionals with_openssl [ openssl ];
  # ,,, snip ...
}

```

</details>

Maintainer(s) MAY choose to add both `with` and `enable` feature flags even if
they map one-to-one for consistency with other packages.

<details><summary>Example</summary>

```
{ lib
, stdenv
, openssl
, with_openssl ? true
, enable_ssl ? with_openssl
}:

assert enable_ssl -> with_openssl;

stdenv.mkDerivation {
  # ... snip ...
  configureFlags = lib.optionals enable_ssl ["--enable-ssl"];
  buildInputs = lib.optionals with_openssl [ openssl ];
  # ,,, snip ...

  passthru.features = {
    inherit with_openssl enable_ssl;
  };
}
```

</details>

The rules are to be enforced by static code analyse linter to be written. Since
no static code analyzis is perfect, it shall have support for inhibiting
warnings in unsual cases.

> All feature parameters SHOULD be exported in `features` passthrough set. See example above.

Reviewer 4: ```suggestion
All feature parameters SHOULD be exported in `passthru.features` set. See example above.
```

stdenv.mkDerivation {
   # ... snip ...

   configureFlags = lib.optionals enable_http ["--enable-http"]
                 ++ lib.optionals enable_ftp ["--enable-ftp"];
   buildInputs = lib.optionals withCurl [ curl ];

   # ... snip ...

   passthru.features = {
      inherit with_curl enable_ftp enable_http;
   };

   # ... snip ...
}
```
</details>

When build dependency comes from particular package set, it MAY make sense to
name feature parameter after it. E.g build dependency on `qt6.qtbase` SHOULD
have `with_qt6` feature parameter. If that results in feature parameter name
that is too generic, package name from the set MAY be included into the feature
parameter name. Optional dependency on `pythonPackages.pillow` MAY have feature
parameter `with_python_pillow`.

Build dependency on bindings to low-level libraries SHOULD be named after
underlying library. For example, optional dependecy on `pyqt5` Python bindings
to `Qt5` library should have `with_qt5` feature parameter.

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

This is how one can query list of feature parameters that incur extra runtime dependencies:

```

> $ nix eval --json --impure --expr 'with import ./. {}; builtins.attrNames gnuplot.override.__functionArgs' | jq '.[]| select(startswith("with"))'

Reviewer 4: I would like if we can do it without jq or similar shell script tricks.

But I believe it would require a too huge modification to the language. Nix does not look like a query language.

Reviewer 3: It's not that people have not implemented typed (and checked) arguments in Nix, they have, multiple times. So the massive change would be on Nixpkgs side… and the speed impact will be unpleasant

   penalty for already compliant packages and packages that has no feature
   parameters, but would be quite verbose and require sophisticated automations
   to introduce and drop.

## Other ways to pass feature parameters

### Group all feature parameters into separate attrset parameter

Instead of writing

```
{ lib, stdenv, enable_foo ? true }:

# uses enable_foo
stdenv.mkDerivation { ... }
```

derivation can be written as
```
{ lib, stdenv, features ? { enable_foo = true; }}:

# uses features.enable_foo
stdenv.mkDerivation { ... }
```

It is definitely looks cleaner, but unfortunately hits several limitations of
the Nix language.

1. Nix language provides way to [introspect function argument names](https://nixos.org/manual/nix/stable/language/builtins.html#builtins-functionArgs),
   but no way to learn their default values. Learning default values is fundamentally impossible, since default values of one parameter might depend
   on the value of another parameter, so default values can't be evaluated until function is called.

2. Overrides become much more verbose. Simple and ubiquitous

   ```
   bar.override { with_foo = true; }
   ```

   becomes unwieldy

   ```

>    bar.override (old: old // { features = old.features // { with_foo = true }; })

Reviewer 4: you mean

```suggestion
   bar.overrideAttrs (old: old // { features = old.features // { with_foo = true }; })
```

?

Author: No, I mean `override`. If we do something like this:

```
{ lib, stdenv, features ? { with_foo = true; enable_bar = stdenv.isLinux; } }:

stdenv.mkDerivation { ... }
```

then we will have nice separation of feature flags and other stuff, but overrides become mess.

incompatible with older installations of the Nix.

## Other ways to name parameters

Other naming conventions were considered and found less appealing.

1. Lisp-style `with-foo` variable names are inconvenient when they need to be
   passed to the builder environment as opposed being used exclusively at
   evaluation phase, since Bash does not support referring to environment
   variables that are not C identifier.

2. Camel case is the most popular naming convention in Nixpkgs at the time of
   writing, but adding prefix results in quite unnatural-looking identifiers
   like `enableSsh` or `withLlvm`. In addition, proposed convention visually
   distinguish feature parameters from other kinds of parameters.

3. Camel case, but allowing to follow upstream spelling, produces naturally
   looking identifies like `enableSSH`, `withLibreSSL`, `enableAudio` that are
   wildly inconsistent between each other.

# Prior art
[prior-art]: #prior-art

As mentioned in motivation part, the best in class of feature flag
configuration system is Gentoo Linux:

- [https://www.gentoo.org/support/use-flags]

There is not much work I can find about this problem in Nixpkgs other than my

> previous attempts to solve it on package-by-package basis:

Reviewer 4: ```suggestion
previous attempts to solve it on a  per package basis:
```

disabled is more complicated and and involves more "you just have to know it"
compared to already existing systems.

1. Writing generic code is complicated. Even if list of packages all have
   optional dependency on package `foo`, duck-typing like

```
map (p: p.override { with_foo = false; }) packages
```

won't work because of lack of uniformity, leading to ugly code downstream.

2. If user wants to globally disable support for optional feature or build
   dependency as much as possible, he needs to set multiple feature parameters,
   and he still can miss some. E.g, it is not obvious whether the following
   list is enough to cover everything X11-related:

```
withX
withX11
x11Support
enableX
```

3. There is no clear separation between feature parameters and other arguments
   to the package. With naming mandated by this RFC, distinction is clear to
   cater to automatic queries.

4. Side effect of the proposdd migration plan is compiling of almost
   exahustive list of known feature parameters.

# Detailed design
[design]: #detailed-design

When upstream build system has options to enable or disable a optional features

> or configure some compile-time parameter, like path to default mailbox or

Reviewer 6: Would this rule apply to feature parameters that do not correspond to upstream build system features?

For exampe, [libiconv](https://github.com/NixOS/nixpkgs/blob/fadc1d7ef25cba8580cc0380f07e686b1ad2fe49/pkgs/development/libraries/libiconv/default.nix#L4) has a feature flag `enableDarwinABICompat` for building libiconv in a way that is ABI-compatible with the system-provided libiconv. The plan is to drop darwin.libiconv in favor of the upstream one, but the upstream one needs this compatibility to allow references to the system one to be replace with it (using `install_name_tool`). Otherwise, doing so will result in crashes due to missing symbols.

Additionally, libiconv (and other packages) feature `enableShared` and `enableStatic` flags. Would these be migrated to the proposed feature flags section or left outside of it?

Author: I'd say to migrate these too, in line with suggested `enable_ssl`.

Reviewer 6: Thanks. My other question was about options that are for nixpkgs use. libiconv is the primary one that comes to mind. It builds libivonv in a specific way for ABI compatibility, so it doesn’t map to a specific build system option.

If I am understanding correctly, that type of option would still be okay under this RFC even though it’s not an upstream-provided option. (In this case, we’re manually recreating what Apple does in their fork’s Xcode project.)

feature flags. Further text will use wording "corresponding feature parameter"
with implicit caveat "provided that package maintainer decided to export
underlying build system configuration option".

Feature parameters MUST only be used according to the rules outlined in this
RFC. They MUST NOT be used for any other purpose.

Boolean feature parameter MUST be either `with`-parameter or `enable`-parameter.
Non-boolean (e.g. string or integer) feature parameter MUST be an `conf`-parameter.

Feature parameter that incurs additional runtime dependency MUST be a `with`-parameter.

<details><summary>Example</summary>

```
{ lib, stdenv, mpfr, with_mpfr ? true }:

stdenv.mkDerivation {
  # ... snip ...
  configureFlags = lib.optionals with_mpfr ["--with-mpfr"];
  buildInputs = lib.optionals with_mpfr [ mpfr ];
  # ... snip ...
}
```

</details>

As special provision, if optional vendored dependency is exposed by feature
parameter, it MUST be `with`-parameter.

If feature parameter does not incur extra runtime dependencies, corresponding
feature parameter MUST be `enable`-parameter and its name SHOULD correspond to

> the upstream naming, unless is causes major inconsistency with other packages.

Reviewer 6: Some projects use build systems that don’t support feature flags but do support customization via environment variables or preprocessor definitions (e.g., makefiles or xcodebuild). Am I correct assuming there be some leeway when translating their flags to appropriate feature flags in nixpkgs?

For example, MoltenVK has an [option](https://github.com/KhronosGroup/MoltenVK?tab=readme-ov-file#metal_private_api) to determine whether to use private APIs. That option is named `MVK_USE_METAL_PRIVATE_API`. A direct translation would be `enable_mvk_use_metal_private_api`, which is verbose and funky. The current package uses [`enablePrivateAPIUsage`](https://github.com/NixOS/nixpkgs/blob/b06025f1533a1e07b6db3e75151caa155d1c7eb3/pkgs/os-specific/darwin/moltenvk/default.nix#L23), which I assume would translate to something like `enable_private_api_usage` or `enable_metal_private_api_usage`.

Author: Yes, consistency across nixpkgs is more important than matching upstream naming exactly (I think I have this line somewhere).

I think we can call this option `enable_private_api`, and use this name for other, similar situations.
I'll add your example into guidelines.

  be updated to mention that feature parameters use snake case convention unlike other variables that use
  `lowerCamelCase'.

### Technical process

Renaming the parameters is handled automatically by the `makeOverrideable`
function, as implemented in [https://github.com/NixOS/nixpkgs/pull/284107](#284107).

Renaming `nonCompliantFoo` into `enable_foo` involves two steps:

 1. Mechanical renaming all mentions of `nonCompliantFoo` in package
    definition.
 2. Adding renaming rule into
    [https://github.com/KAction/nixpkgs/blob/contrib/0/rfc169-proof/out/lib/rfc0169.json](makeOverrideable)
    configuration, if it is not there already.

## Feature parameter naming guidelines

Rules in the first section of this RFC describe different scenarios when
`with`, `enable` and `conf` feature parameters must be used, but coming with
exact name that is clear, concise, describes it effect correctly and is
consistent across the nixpkgs is hard to codify with absolute rule. Instead,
this RFC provides set of guidelines and examples based on scenarios brought up
during the discussion.

Feature flags that require single external dependency SHOULD be named after

> that dependency. Prefix `lib` SHOULD be removed. For example,

Author: I don't like this change. Why?

> that dependency. Prefix `lib` SHOULD be removed. For example,

Author: Otherwise we will have `with_alsalib` and `with_libgif`.

<details><summary>Example</summary>

```
systemd => with_systemd
libgif  => with_gif
curl    => with_curl

> alsa    => with_alsa

Author: No. There is rule about removing `lib` prefix.

```

## Other ways to do migration

1. Put deprecation logic into individual packages. That would avoid evaluation
   penalty for already compliant packages and packages that has no feature
   parameters, but would be quite verbose and require sophisticated automations
   to introduce and drop deprecation logic.

## Other ways to pass feature parameters

### Group all feature parameters into separate attrset parameter

Instead of writing

```
{ lib, stdenv, enable_foo ? true }:

# uses enable_foo
stdenv.mkDerivation { ... }
```

derivation can be written as
```
{ lib, stdenv, features ? { enable_foo = true; }}:

# uses features.enable_foo
stdenv.mkDerivation { ... }
```

It is definitely looks cleaner, but unfortunately hits several limitations of
the Nix language.

1. Nix language provides way to [introspect function argument names](https://nixos.org/manual/nix/stable/language/builtins.html#builtins-functionArgs),
   but no way to learn their default values. Learning default values is fundamentally impossible, since default values of one parameter might depend
   on the value of another parameter, so default values can't be evaluated until function is called.

> 2. Overrides become much more verbose. Simple and ubiquitous

Reviewer 5: I've had a little breakthrough regarding this: https://github.com/NixOS/nixpkgs/pull/312432
